rir/src/interpreter/interp.cpp

```diff
@@ -536,82 +536,76 @@ INLINE SEXP fixupAST(SEXP call, Context* ctx, size_t nargs) {
+           TYPEOF(callee) == CLOSXP);
+
     CallSite* cs = caller->callSite(id);
     SEXP call = cp_pool_at(ctx, cs->call);
 
     profileCall(cs, callee);
 
-    Protect p;
-    if (argsOnStack)
-        if (TYPEOF(callee) == SPECIALSXP || TYPEOF(callee) == CLOSXP) {
+    if (TYPEOF(callee) == SPECIALSXP) {
+        assert(call != R_NilValue);
+        Protect p;
+        if (argsOnStack) {
             call = fixupAST(call, ctx, nargs);
             p(call);
-        }
-
-    SEXP result;
-
-    switch (TYPEOF(callee)) {
-    case SPECIALSXP: {
-        assert(call != R_NilValue);
-        if (argsOnStack)
             ostack_popn(ctx, nargs);
+        }
         // get the ccode
         CCODE f = getBuiltin(callee);
         int flag = getFlag(callee);
         R_Visible = static_cast<Rboolean>(flag != 1);
         warnSpecial(callee, call);
         // call it with the AST only
-        result = f(call, callee, CDR(call), env);
+        SEXP result = f(call, callee, CDR(call), env);
         if (flag < 2)
             R_Visible = static_cast<Rboolean>(flag != 1);
-        break;
+        return result;
     }
-    case BUILTINSXP: {
-        // create the argslist
-        SEXP argslist;
-        if (argsOnStack) {
-            argslist = createArgsListStack(caller, nargs, cs, env, ctx, true);
-            ostack_popn(ctx, nargs);
-        } else {
-            argslist = createArgsList(caller, call, nargs, cs, env, ctx, true);
-        }
-        p(argslist);
+
+    // create the argslist
+    bool eager = TYPEOF(callee) == BUILTINSXP;
+    SEXP argslist;
+    Protect p;
+    if (argsOnStack) {
+        call = fixupAST(call, ctx, nargs);
+        p(call);
+        argslist = createArgsListStack(caller, nargs, cs, env, ctx, eager);
+        ostack_popn(ctx, nargs);
+    } else {
+        argslist = createArgsList(caller, call, nargs, cs, env, ctx, eager);
+    }
+    p(argslist);
+
+    if (TYPEOF(callee) == BUILTINSXP) {
         // get the ccode
         CCODE f = getBuiltin(callee);
         int flag = getFlag(callee);
         if (flag < 2)
             R_Visible = static_cast<Rboolean>(flag != 1);
         // call it
-        result = f(call, callee, argslist, env);
+        SEXP result = f(call, callee, argslist, env);
         if (flag < 2)
             R_Visible = static_cast<Rboolean>(flag != 1);
-        break;
+        return result;
     }
-    case CLOSXP: {
-        // create the argslist
-        SEXP argslist;
-        if (argsOnStack) {
-            argslist = createArgsListStack(caller, nargs, cs, env, ctx, false);
-            ostack_popn(ctx, nargs);
-        } else {
-            argslist = createArgsList(caller, call, nargs, cs, env, ctx, false);
-        }
-        p(argslist);
+
+    if (TYPEOF(callee) == CLOSXP) {
         // if body is EXTERNALSXP, it is rir serialized code, execute it
         // directly
-        SEXP body = BODY(callee);
+        SEXP body = BODY(callee), result;
```

Comment:
So it's not immediately obvious to me why you're using the comma operator. Is this a typo, or am I just confused?

Reply:
Oh, this is just a declaration... I agree that it's kind of ugly